src/storage/src/table/state_table.rs

```diff
@@ -168,83 +169,114 @@ impl<'a, S: StateStore> StateTableRowIter<'a, S> {
-                        next_flag = NextOutcome::MemTable;
-
-                        match mem_table_row_op {
-                            RowOp::Insert(row) => res = Some(row.clone()),
-                            RowOp::Delete(_) => {}
-                            RowOp::Update(_) => {
-                                panic!(
+                        Ordering::Greater => {
+                            // mem_table_item will be return
+
+                            match mem_table_row_op {
+                                RowOp::Insert(row) => {
+                                    res = Some(row.clone());
+                                    next_flag = NextOutcome::MemTable;
+                                }
+                                RowOp::Delete(_) => {
+                                    next_flag = NextOutcome::Pass;
+                                }
+                                RowOp::Update(_) => {
+                                    panic!(
                                     "There must be a record in shared storage, so this case is unreachable.",
                                 );
+                                }
                             }
+                            self.cell_based_item = Some((cell_based_pk, cell_based_row));
                         }
-                        self.cell_based_item = Some((cell_based_pk, cell_based_row));
                     }
                 }
             }
-        }
-        match next_flag {
-            NextOutcome::MemTable => {
-                self.mem_table_iter.next();
-            }
-            NextOutcome::Storage => {
-                self.cell_based_item = self.cell_based_streaming_iter.next().await.map_err(err)?;
-            }
-            NextOutcome::Both => {
-                self.mem_table_iter.next();
-                self.cell_based_item = self.cell_based_streaming_iter.next().await.map_err(err)?
+
+            match next_flag {
+                NextOutcome::MemTable => {
+                    self.mem_table_iter.next();
+                    break;
+                }
+                NextOutcome::Storage => {
+                    self.cell_based_item =
+                        self.cell_based_streaming_iter.next().await.map_err(err)?;
+                    break;
+                }
+                NextOutcome::Both => {
+                    self.mem_table_iter.next();
+                    self.cell_based_item =
+                        self.cell_based_streaming_iter.next().await.map_err(err)?;
+                    break;
+                }
+                NextOutcome::End => {
+                    break;
+                }
+                NextOutcome::Pass => {
+                    // if a pk exist in both shared storage(cell_based_table) and
+                    // memory(mem_table), and
+                    // mem_table stores a delete record, just next again
+                    self.mem_table_iter.next();
```

Comment:
Did you fuse both iterators? What's the expected outcome if any iterator reached the end but still calling `.next`?